### PR TITLE
[Tune] Don't raise TuneError on `fail_fast="raise"`

### LINKED ISSUE #30816: [AIR/Tune] `fail_fast="raise"` does not raise the actual exception in Tuner
In `Tuner` we are always raising `TuneError`, and never the actual exception from the Trainable. This is not the behavior that is mentioned in the docstring.

These tests fail
```
   def test_tuner_fail_fast_true(self):
        trainer = FailingTrainer()
        param_space = {
            "scaling_config": ScalingConfig(num_workers=tune.grid_search([1, 2]))
        }

        failure_config = FailureConfig(fail_fast=True)

        tuner = Tuner(
            trainable=trainer,
            run_config=RunConfig(
                name="test_tuner_trainer_fail", failure_config=failure_config
            ),
            param_space=param_space,
            tune_config=TuneConfig(mode="max", metric="iteration"),
        )
        with self.assertRaises(TuneError):
            results = tuner.fit()
            errors = [result.error for result in results if result.error]
            assert len(errors) == 1

    def test_tuner_fail_fast_raise(self):
        trainer = FailingTrainer()
        param_space = {
            "scaling_config": ScalingConfig(num_workers=tune.grid_search([1, 2]))
        }

        failure_config = FailureConfig(fail_fast="raise")

        tuner = Tuner(
            trainable=trainer,
            run_config=RunConfig(
                name="test_tuner_trainer_fail", failure_config=failure_config
            ),
            param_space=param_space,
            tune_config=TuneConfig(mode="max", metric="iteration"),
        )
        with self.assertRaises(RuntimeError):
            tuner.fit()
```

_Originally posted by @amogkam in https://github.com/ray-project/ray/issues/30813#issuecomment-1333005237_

### PR DESCRIPTION
**Pull request by amogkam**
As the docstring mentioned, `fail_fast="raise"` should raise the exception received by the Trainable. This is the behavior for `tune.run`, but in `Tuner` API, a generic `TuneError` is raised.

This PR changes the behavior of Tuner to match with tune.run. 

This PR also makes sure that errors raised in the **driver** (from Callbacks, or input validation before the experiment starts) are raised directly and not wrapped in `TuneError`, matching the behavior with `tune.run`

Closes #30816 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/tuner.py [hunk 1]
```diff
@@ -270,7 +270,7 @@ def fit(self) -> ResultGrid:
         if not self._is_ray_client:
             try:
                 return self._local_tuner.fit()
-            except Exception as e:
+            except TuneError as e:
```
Comment (matthewdeng) [inline python/ray/tune/tuner.py:273]: Does this affect any other cases? I.e. are there other paths that are not raising a `TuneError` but are expected to be intercepted here?

**Review by matthewdeng — APPROVED** [review]
One question regarding error handling but otherwise LGTM.

File: python/ray/tune/tuner.py (hunk 1 above)
Comment (amogkam) [reply python/ray/tune/tuner.py:273]: Not sure. I'm hoping tests will cover this if there are any. @xwjiang2010 do you know?

Comment (xwjiang2010) [reply python/ray/tune/tuner.py:273]: Not sure either but looks like some tests are failing. Maybe start from there?

File: python/ray/tune/tuner.py [hunk 2]
```diff
@@ -292,7 +292,7 @@ def fit(self) -> ResultGrid:
                     string_queue,
                 )
                 return ray.get(fit_future)
-            except Exception as e:
+            except TuneError as e:
```
Comment (xwjiang2010) [inline python/ray/tune/tuner.py:295]: Just making sure that ray.get() preserves the original type of error right?

**Comment by xwjiang2010** [comment]
hmmm, test_tuner is still failing.

**Review by Yard1 — APPROVED** [review]

**Review by xwjiang2010 — APPROVED** [review]

**Pull request merged**